### ISSUE TITLE
Nan loss during training

### ISSUE THREAD

**Issue by jamesguoxin** [google/bi-tempered-loss#13]
Dear authors,

I used bi_tempered_binary_logistic_loss as my NN's loss function, and I noticed that after a few hundreds steps training, some loss becomes nan. I printed the value outputted by bi_tempered_binary_logistic_loss and confirmed the nan value was from bi_tempered_binary_logistic_loss. Any idea why? Do I need to adjust the parameter setting and how? Thank you for answering!!

**Comment by eamid** [comment]
Can you please share more details about your hyperparameters? Generally, if your model is stable with `t1 = t2 = 1.0` (i.e. cross entropy loss), it should also run without a problem when `t1 < 1.0` and `t2 > 1.0`.

**Comment by jamesguoxin** [comment]
Currently I'm using t1=0.8 and t2=2.0. I printed value of probabilities and loss_value in _custom_gradient_bi_tempered_logistic_loss(activations) function, and I didn't find any nan in probabilities, but identified [-nan 6.30556154] in loss_value. This value caused nan propagation in loss.

**Comment by eamid** [comment]
Can you please run the same model using `t1 = t2 = 1.0`?

**Comment by jamesguoxin** [comment]
Hi @eamid 

I switched to t1 = t2 = 1.0, the training seems OK, nan value doesn't appear

**Comment by eamid** [comment]
Thanks for checking! This is a bit strange since we expect bi-tempered loss to be more stable than cross entropy. If you can, please also try `t1=0.8, t2=1.0` and `t1=1.0, t2=2.0` to see which component is causing the instability. You can also try temperature pairs closer to 1.0, e.g. `t1=0.99, t2=1.01`. I am happy to take a look at your code if you can share it here.

**Comment by jamesguoxin** [comment]
Hi @eamid 

I tried with following settings
`t1=0.8, t2=1.0`
`t1=1.0, t2=2.0`
`t1=0.99, t2=1.01`

All of them produced nan loss after a few steps training. I could introduce our scenario and some observations first. We would like to utilize bi-tempered loss to boost our recommendation model, as for training data, about 5% of them is positive samples (clicked), rest is negative. Our model's output is simply a logit, so I use `bi_tempered_binary_logistic_loss` for experiment. The code is the same as your open-sourced version. 

```python
def bi_tempered_binary_logistic_loss(activations,
                                     labels,
                                     t1,
                                     t2,
                                     label_smoothing=0.0,
                                     num_iters=5):
  with tf.name_scope('binary_bitempered_logistic'):
    t1 = tf.convert_to_tensor(t1)
    t2 = tf.convert_to_tensor(t2)
    out_shape = tf.shape(labels)
    labels_2d = tf.reshape(labels, [-1, 1])
    activations_2d = tf.reshape(activations, [-1, 1])
    internal_labels = tf.concat([1.0 - labels_2d, labels_2d], 1)
    internal_logits = tf.concat([tf.zeros_like(activations_2d), activations_2d],
                                1)
    losses = bi_tempered_logistic_loss(internal_logits, internal_labels, t1, t2,
                                       label_smoothing, num_iters)
    return tf.reshape(losses, out_shape)


def bi_tempered_logistic_loss(activations,
                              labels,
                              t1,
                              t2,
                              label_smoothing=0.0,
                              num_iters=5):
  with tf.name_scope('bitempered_logistic'):
    t1 = tf.convert_to_tensor(t1)
    t2 = tf.convert_to_tensor(t2)
    if label_smoothing > 0.0:
      num_classes = tf.cast(tf.shape(labels)[-1], tf.float32)
      labels = (
          1 - num_classes /
          (num_classes - 1) * label_smoothing) * labels + label_smoothing / (
              num_classes - 1)

    @tf.custom_gradient
    def _custom_gradient_bi_tempered_logistic_loss(activations):
      with tf.name_scope('gradient_bitempered_logistic'):
        probabilities = tempered_softmax(activations, t2, num_iters)
        
        loss_values = tf.multiply(
            labels,
            log_t(labels + 1e-10, t1) -
            log_t(probabilities, t1)) - 1.0 / (2.0 - t1) * (
                tf.pow(labels, 2.0 - t1) - tf.pow(probabilities, 2.0 - t1))

        def grad(d_loss):
          delta_probs = probabilities - labels
          forget_factor = tf.pow(probabilities, t2 - t1)
          delta_probs_times_forget_factor = tf.multiply(delta_probs,
                                                        forget_factor)
          delta_forget_sum = tf.reduce_sum(
              delta_probs_times_forget_factor, -1, True)
          escorts = tf.pow(probabilities, t2)
          escorts = escorts / tf.reduce_sum(escorts, -1, True)
          derivative = delta_probs_times_forget_factor - tf.multiply(
              escorts, delta_forget_sum)
          return tf.multiply(d_loss, derivative)

        return loss_values, grad

    def reduced_loss(activations):
      return tf.reduce_sum(
          _custom_gradient_bi_tempered_logistic_loss(activations), -1)

    loss_values = tf.cond(
        tf.math.logical_and(tf.equal(t1, 1.0), tf.equal(t2, 1.0)),
        functools.partial(
            tf.nn.softmax_cross_entropy_with_logits,
            labels=labels,
            logits=activations), functools.partial(reduced_loss, activations))
    return loss_values
```

I printed the `probabilities` fed into the `loss_values` calculation in `_custom_gradient_bi_tempered_logistic_loss ` function and identified which probabilities value caused nan. It seems [0.996272922 0.0037270959] -> [-nan 6.30556154]

**Comment by jamesguoxin** [comment]
One more thing to note, we used `tf.losses.sigmoid_cross_entropy` as loss function before. And `t1=1.0, t2=1.0` works fine in our case.

**Comment by eamid** [comment]
Thanks for the info. I tried the loss with `t1=0.8, t2=2.0` and a sigmoid probability of `0.996272922`. At least in this setting, I am not getting a NaN. Can you please also share the logit values and the labels that are causing a NaN?

**Comment by jamesguoxin** [comment]
Hi @eamid 

After careful debugging, I think the issue might be due to our side. I print internal_logits and internal_labels as you suggested, I share the value caused NaN here

`internal_logits = [0 -2.14406967]`
`internal_labels = [-1 2]`
`probabilities = [0.895113289 0.104886681]`
`loss_values = [-nan 3.25809574]`

The internal_labels value seems wired. I will look into the training data further and update this issue if I found something.

**Comment by eamid** [comment]
Your `internal_labels = [-1 2]` seem wrong. The labels should be in between 0 and 1 (i.e. correspond to probabilities).
